File: android/framework_old/src/main/java/com/dreampany/network/manager/NetworkManager.java
/*
package com.dreampany.network.manager;

import android.content.Context;

import androidx.annotation.Nullable;

import com.dreampany.network.api.BluetoothApi;
import com.dreampany.network.api.InternetApi;
import com.dreampany.network.api.MobileApi;
import com.dreampany.network.api.WifiApApi;
import com.dreampany.network.api.WifiApi;
import com.dreampany.network.data.model.Network;
import com.dreampany.network.misc.RxMapper;
import com.google.common.collect.Maps;
import com.google.common.collect.Sets;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;

import javax.inject.Inject;
import javax.inject.Singleton;

*/
/**
 * Created by Hawladar Roman on 8/18/2018.
 * Dreampany Ltd
 * dreampanymail@gmail.com
 *//*

@Singleton
public final class NetworkManager {

    public interface Callback {
        void onResult(Network... networks);
    }

    private final Context context;
    private final RxMapper rx;
*/
/*    private final WifiApi wifi;
    private final WifiApApi ap;
    private final MobileApi mobile;
    private final BluetoothApi bt;
    private final InternetApi internetApi;*//*

    private volatile boolean internet;
    private final Set<Callback> callbacks;
    private final Map<Callback, Boolean> checkInternets;

    private boolean resultFired;

    @Inject
    NetworkManager(Context context,
                   RxMapper rx,
                   WifiApi wifi,
                   WifiApApi ap,
                   MobileApi mobile,
                   BluetoothApi bt,
                   InternetApi internetApi) {
        this.context = context.getApplicationContext();
        this.rx = rx;
        this.wifi = wifi;
        this.ap = ap;
        this.mobile = mobile;
        this.bt = bt;
        this.internetApi = internetApi;
        callbacks = Sets.newConcurrentHashSet();
        checkInternets = Maps.newConcurrentMap();

        resultFired = false;
    }

    void onResult(boolean internet) {
        resultFired = true;
        this.internet = internet;
        postActiveNetworks();
    }

    public void observe(Callback callback, boolean checkInternet) {
        callbacks.add(callback);
        checkInternets.put(callback, checkInternet);
        startInternetIfPossible();
    }

    public void deObserve(Callback callback, boolean stopInternetCheck) {
        callbacks.remove(callback);
        if (stopInternetCheck) {
            checkInternets.remove(callback);
        }
        stopInternetIfPossible();
    }

    public boolean isObserving() {
        return resultFired && !callbacks.isEmpty();
    }

    public List<Network> getNetworks() {
        List<Network> networks = new ArrayList<>();
        networks.add(wifi.getNetwork(internet));
        networks.add(ap.getNetwork(internet));
        networks.add(bt.getNetwork(internet));
        return networks;
    }

    public List<Network> getActiveNetworks() {
        List<Network> networks = new ArrayList<>();
        if (wifi.isEnabled()) {
            networks.add(wifi.getNetwork(internet));
        }
        if (ap.isEnabled()) {
            networks.add(ap.getNetwork(internet));
        }
        if (mobile.isEnabled()) {
            networks.add(mobile.getNetwork(internet));
        }
*/
/*        if (bt.isEnabled()) {
            networks.add(bt.getNetwork(internet));
        }*//*

        return networks;
    }

    public boolean hasInternet() {
        return internet;
    }

    private void startInternetIfPossible() {
        for (Map.Entry<Callback, Boolean> entry : checkInternets.entrySet()) {
            if (entry.getValue()) {
                internetApi.start(this::onResult);
                break;
            }
        }
    }

    private void stopInternetIfPossible() {
        boolean checkInternet = false;
        for (Map.Entry<Callback, Boolean> entry : checkInternets.entrySet()) {
            if (entry.getValue()) {
                checkInternet = true;
                break;
            }
        }
        if (!checkInternet) {
            internetApi.stop(this::onResult);
        }
    }

    private void postActiveNetworks() {
        List<Network> result = getActiveNetworks();
        Network[] networks = result.toArray(new Network[0]);
        for (Callback callback : callbacks) {
            callback.onResult(networks);
        }
    }
}
*/

File: android/framework_old/src/main/kotlin/com/dreampany/network/manager/NetworkManager.kt
package com.dreampany.network.manager

import android.content.Context
import com.dreampany.framework.misc.AppExecutor
import com.dreampany.framework.misc.RxMapper
import com.dreampany.network.api.Wifi
import com.dreampany.network.data.model.Network
import com.github.pwittchen.reactivenetwork.library.rx2.Connectivity
import com.github.pwittchen.reactivenetwork.library.rx2.ReactiveNetwork
import io.reactivex.Single
import io.reactivex.disposables.Disposable
import timber.log.Timber
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Created by Roman-372 on 7/1/2019
 * Copyright (c) 2019 bjit. All rights reserved.
 * hawladar.roman@bjitgroup.com
 * Last modified $file.lastModified
 */
@Singleton
class NetworkManager
@Inject constructor(
    private val context: Context,
    private val rx: RxMapper,
    private val ex: AppExecutor,
    private val wifi: Wifi
) {

    interface Callback {
        fun onNetworks(networks: List<Network>)
    }

    private var disposable: Disposable? = null
    private val callbacks = mutableSetOf<Callback>()
    private val networks = mutableListOf<Network>()
    private var checkInternet: Boolean = false

    fun observe(callback: Callback, checkInternet: Boolean = false) {
        callbacks.add(callback)
        this.checkInternet = checkInternet
        if (isStarted()) {
            ex.postToUi(Runnable {
                postNetworks(callback)
            })
            return
        }
        disposable = ReactiveNetwork.observeNetworkConnectivity(context)
            .flatMapSingle({ buildNetwork(it) })
            .subscribeOn(rx.io())
            .observeOn(rx.io())
            .subscribe(this::postResult, this::postError)
    }

    fun deObserve(callback: Callback) {
        callbacks.remove(callback)
        if (callbacks.isEmpty()) {
            if (isStarted()) {
                disposable?.dispose()
            }
        }
    }

    fun hasInternet(): Boolean {
        for (network in networks) {
            if (network.internet) {
                return true
            }
        }
        return false
    }

    fun isObserving(): Boolean {
        return !callbacks.isEmpty();
    }

    private fun isStarted(): Boolean {
        disposable?.let {
            if (it.isDisposed) {
                return false
            }
            return true
        }
        return false
    }

    private fun postNetworks(callback: Callback) {
        if (!networks.isEmpty()) {
            callback.onNetworks(networks)
        }
    }

    private fun postNetworks() {
        for (callback in callbacks) {
            postNetworks(callback)
        }
    }

    private fun postResult(network: Network) {
        Timber.v("postReesult %s", network.toString())
        if (networks.contains(network)) {
            networks.set(networks.indexOf(network), network)
        } else {
            networks.add(network)
        }
        postNetworks()
    }

    private fun postError(error: Throwable) {
        postNetworks()
    }

    private fun buildNetwork(connectivity: Connectivity): Single<Network> {
        Timber.v(
            "Connectivity %s %s %s",
            connectivity.typeName(),
            connectivity.available(),
            connectivity.toString()
        )
        return Single.create({
            val network = wifi.getNetwork()
            if (connectivity.available()) {
                if (checkInternet) {
                    network.internet = ReactiveNetwork.checkInternetConnectivity().blockingGet()
                }
            }
            it.onSuccess(network)
        })
    }
}